tower-http/src/trace.rs

```diff
@@ -0,0 +1,348 @@
+    pub fn new() -> Self {
+        Self {
+            latency_unit: LatencyUnit::Millis,
+            classifier: DefaultHttpResponseClassifier::default(),
+            event_emitter: DefaultEmitTracingEvents::default(),
+        }
+    }
+}
+
+impl<Classifier, EventEmitter> TraceLayer<Classifier, EventEmitter> {
+    /// Change the latency unit events will use.
+    #[inline]
+    pub fn latency_unit(mut self, latency_unit: LatencyUnit) -> Self {
+        self.latency_unit = latency_unit;
+        self
+    }
+
+    /// Change the [`ClassifyResponse`] that will be used.
+    ///
+    /// [`ClassifyResponse`]: crate::classify_response::ClassifyResponse
+    #[inline]
+    pub fn classify_responses_with<NewClassifier>(
+        self,
+        classifier: NewClassifier,
+    ) -> TraceLayer<NewClassifier, EventEmitter> {
+        TraceLayer {
+            classifier,
+            latency_unit: self.latency_unit,
+            event_emitter: self.event_emitter,
+        }
+    }
+
+    /// Change the [`EmitTracingEvents`] that will be used.
+    #[inline]
+    pub fn emit_events_with<NewEventEmitter>(
+        self,
+        event_emitter: NewEventEmitter,
+    ) -> TraceLayer<Classifier, NewEventEmitter> {
+        TraceLayer {
+            event_emitter,
+            classifier: self.classifier,
+            latency_unit: self.latency_unit,
+        }
+    }
+}
+
+impl<Classifier> TraceLayer<Classifier, DefaultEmitTracingEvents> {
+    /// Change whether headers should be recorded on the span.
+    ///
+    /// Defaults to `false`.
+    ///
+    /// Note that this method is only accessible if you're using [`DefaultEmitTracingEvents`]
+    #[inline]
+    pub fn record_headers(mut self, record_headers: bool) -> Self {
+        self.event_emitter.record_headers = record_headers;
+        self
+    }
+
+    /// Change whether the full URI should be recorded on the span or if only the path should be
+    /// recorded.
+    ///
+    /// Defaults to only recording the path no the full URI.
+    ///
+    /// Note that this method is only accessible if you're using [`DefaultEmitTracingEvents`]
+    #[inline]
+    pub fn record_full_uri(mut self, record_full_uri: bool) -> Self {
+        self.event_emitter.record_full_uri = record_full_uri;
+        self
+    }
+}
+
+impl<S, Classifier, EventEmitter> Layer<S> for TraceLayer<Classifier, EventEmitter>
+where
+    Classifier: Clone,
+    EventEmitter: Clone,
+{
+    type Service = Trace<S, Classifier, EventEmitter>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        Trace {
+            inner,
+            latency_unit: self.latency_unit,
+            classifier: self.classifier.clone(),
+            event_emitter: self.event_emitter.clone(),
+        }
+    }
+}
+
+/// Middleware that add high level [tracing] to a [`Service`].
+///
+/// [tracing]: https://crates.io/crates/tracing
+/// [`Service`]: tower_service::Service
+#[derive(Clone, Debug)]
+pub struct Trace<S, Classifier, EventEmitter> {
```

Comment:
nit: since the classifier and event emitter have defaults on the layer type, they probably should here as well?